schema/encoding/jsonschema.go

```diff
@@ -0,0 +1,169 @@
+		return empty, nil
+	}
+	b := new(bytes.Buffer)
+	switch t := v.(type) {
+	case *schema.String:
+		if _, err = fmt.Fprintf(b, "%q: %q, ", "type", "string"); err != nil {
+			return empty, err
+		}
+		if t.Regexp != "" {
+			if _, err = fmt.Fprintf(b, "%q: %q, ", "pattern", t.Regexp); err != nil {
+				return empty, err
+			}
+		}
+		if len(t.Allowed) > 0 {
+			var allowed []string
+			for _, value := range t.Allowed {
+				allowed = append(allowed, fmt.Sprintf("%q", value))
+			}
+			if _, err = fmt.Fprintf(b, "%q: [ %s ],", "enum", strings.Join(allowed, ",")); err != nil {
+				return empty, err
+			}
+		}
+		if t.MinLen > 0 {
+			if _, err = fmt.Fprintf(b, "%q: %s, ", "minLength", strconv.FormatInt(int64(t.MinLen), 10)); err != nil {
+				return empty, err
+			}
+		}
+		if t.MaxLen > 0 {
+			if _, err = fmt.Fprintf(b, "%q: %s, ", "maxLength", strconv.FormatInt(int64(t.MaxLen), 10)); err != nil {
+				return empty, err
+			}
+		}
+	case *schema.Integer:
+		if _, err = fmt.Fprintf(b, "%q: %q, ", "type", "number"); err != nil {
+			return empty, err
+		}
+
+		if len(t.Allowed) > 0 {
+			var allowed []string
+			for _, value := range t.Allowed {
+				allowed = append(allowed, strconv.FormatInt(int64(value), 10))
+			}
+			if _, err = fmt.Fprintf(b, "%q: [ %s ],", "enum", strings.Join(allowed, ",")); err != nil {
+				return empty, err
+			}
+		}
+		bound, err := boundariesToJSONSchema(t.Boundaries)
+		if err != nil {
+			return empty, err
+		}
+		if _, err = b.Write(bound); err != nil {
+			return empty, err
+		}
+		if _, err = b.WriteString(","); err != nil {
+			return empty, err
+		}
+	case *schema.Array:
+		if _, err = fmt.Fprintf(b, "%q: %q, ", "type", "array"); err != nil {
+			return empty, err
+		}
+		if t.ValuesValidator != nil {
+			v, err := validatorToJSONSchema(t.ValuesValidator)
+			if err != nil {
+				return empty, err
+			}
+			if _, err = fmt.Fprintf(b, "%q: %s, ", "items", v); err != nil {
+				return empty, err
+			}
+		}
+	case *schema.Object:
+		buf, _ := SchemaToJSONSchema(t.Schema)
+		if _, err = fmt.Fprintf(b, "%s, ", buf); err != nil {
+			return empty, err
+		}
+	case *schema.Time:
+		if _, err = fmt.Fprintf(b, "%q: %q, ", "type", "string"); err != nil {
+			return empty, err
+		}
+		if _, err = fmt.Fprintf(b, "%q: %q, ", "format", "date-time"); err != nil {
+			return empty, err
+		}
+	default:
+	}
+	b = bytes.NewBuffer(bytes.TrimRight(b.Bytes(), ", "))
+	return b.Bytes(), nil
+}
+
+var (
+	empty = []byte{}
+)
+
+// SchemaToJSONSchema helper
+func SchemaToJSONSchema(s *schema.Schema) (_ []byte, err error) {
+	b := new(bytes.Buffer)
+	b.WriteString("{")
+	if s.Description != "" {
+		if _, err = fmt.Fprintf(b, "%q: %q, ", "title", s.Description); err != nil {
+			return empty, err
+		}
+	}
+	if _, err = fmt.Fprint(b, `"type": "object",`); err != nil {
+		return empty, err
+	}
+	if _, err = fmt.Fprintf(b, `"properties": {`); err != nil {
+		return empty, err
+	}
+	var required []string
+	fbuf := new(bytes.Buffer)
```

Comment:
Reading quickly thru the code, I don't understand why you need the two buffers `b` and `fbuf`. It looks like you are just appending content of `fbuf` to `b` on each loop iteration. Why couldn't you just write directly into `b` inside the loop?

Also, why don't you pass around this buffer to `validatorToJSONSchema` and `boundariesToJSON`? You should also have a private version of `schemaToJSON` that take this buffer as well.

Did I miss something?